From da53b89bd39c9ce199d6d165d1a7c4d4e464a499 Mon Sep 17 00:00:00 2001
From: CaoSong <caosong@mprc.pku.edu.cn>
Date: Mon, 12 Dec 2011 15:12:11 +0800
Subject: [PATCH 157/641] UniCore64: Add setup_arch_cpuinfo function

Add the function setup_arch_cpuinfo in the kernel/cpu-uc64.c and
the needed macro for it in the asm/elf.h to show the cpu information.
Signed-off-by: Cao Song <caosong@mprc.pku.edu.cn>
---
 arch/unicore64/include/arch/hwdef-copro.h |    6 +++--
 arch/unicore64/include/asm/elf.h          |    2 +-
 arch/unicore64/kernel/cpu-uc64.c          |   35 ++++++++++++++++++++++++-----
 3 files changed, 35 insertions(+), 8 deletions(-)

diff --git a/arch/unicore64/include/arch/hwdef-copro.h b/arch/unicore64/include/arch/hwdef-copro.h
index f51acd4..db91f27 100644
--- a/arch/unicore64/include/arch/hwdef-copro.h
+++ b/arch/unicore64/include/arch/hwdef-copro.h
@@ -87,8 +87,10 @@
  */
 #define CP0_CPUID_PARTNO_MASK		BFMASK(16, 0)
 #define CP0_CPUID_PARTNO_PKUNITY	BFIELD(0x863, 12, 0)
-#define CP0_CPUID_LAYOUT_MASK		BFMASK(4, 16)
-#define CP0_CPUID_SERIES_MASK		BFMASK(4, 20)
+#define CP0_CPUID_LAYOUT_POS		(16)
+#define CP0_CPUID_LAYOUT_MASK		BFMASK(4, CP0_CPUID_LAYOUT_POS)
+#define CP0_CPUID_SERIES_POS		(20)
+#define CP0_CPUID_SERIES_MASK		BFMASK(4, CP0_CPUID_SERIES_POS)
 #define CP0_CPUID_DESIGNER_MASK		BFMASK(8, 24)
 #define CP0_CPUID_DESIGNER_MPRC		BFIELD(0x4D, 8, 24)
 
diff --git a/arch/unicore64/include/asm/elf.h b/arch/unicore64/include/asm/elf.h
index 4341b28..15517e1 100644
--- a/arch/unicore64/include/asm/elf.h
+++ b/arch/unicore64/include/asm/elf.h
@@ -12,7 +12,7 @@ typedef elf_greg_t elf_gregset_t[ELF_NGREG];
 #define ELF_CLASS		ELFCLASS64
 #define ELF_DATA		ELFDATA2LSB
 #define ELF_ARCH		EM_UNICORE64
-#define ELF_PLATFORM		(NULL)
+#define ELF_PLATFORM		"uc64"
 #define ELF_EXEC_PAGESIZE	PAGE_SIZE
 
 #define FP_REGS_NUMBER		33
diff --git a/arch/unicore64/kernel/cpu-uc64.c b/arch/unicore64/kernel/cpu-uc64.c
index b84d23c..fba6a0a 100644
--- a/arch/unicore64/kernel/cpu-uc64.c
+++ b/arch/unicore64/kernel/cpu-uc64.c
@@ -1,11 +1,9 @@
 #include <linux/kernel.h>
 #include <linux/seq_file.h>
 
-void __init setup_arch_cpuinfo(void)
-{
-	/* FIXME */
-	BUG();
-}
+#include <asm/setup_arch.h>
+
+#include <arch/hwdef-copro.h>
 
 /**
  * cpu_uc64_show() - show the UniCore64 cpu information
@@ -89,3 +87,30 @@ const struct seq_operations cpuinfo_op = {
 	.stop	= cpu_uc64_stop,
 	.show	= cpu_uc64_show
 };
+
+/**
+ * setup_arch_cpuinfo() - show the cpu information
+ *
+ * show the cpu information in the screen.
+ * first, get the value in copro through the macro UC64_CPUID,
+ * then check the bit field. The value should be 0x4Duv0863. It represent
+ * that designer is 0x4D, series is 0xu, layout is 0xv, and
+ * part number is 0x0863. And the layout value is not checked here.
+ */
+void __init setup_arch_cpuinfo(void)
+{
+	unsigned long uc64_cpuid;
+
+	uc64_cpuid = UC64_CPUID;
+
+	BUG_ON((uc64_cpuid & CP0_CPUID_PARTNO_MASK) !=
+			CP0_CPUID_PARTNO_PKUNITY);
+
+	BUG_ON((uc64_cpuid & CP0_CPUID_DESIGNER_MASK) !=
+			CP0_CPUID_DESIGNER_MPRC);
+
+	pr_info("CPU: UniCore64, Designer: MPRC, SoC: PKUnity\n");
+	pr_info("revision: %ld, layout: %ld\n",
+		((uc64_cpuid & CP0_CPUID_SERIES_MASK) >> CP0_CPUID_SERIES_POS),
+		((uc64_cpuid & CP0_CPUID_LAYOUT_MASK) >> CP0_CPUID_LAYOUT_POS));
+}
-- 
1.7.9.5

